File: Embarcadero/Berlin/CPP/ThingPoint IoT Sample Data Demo/ThingpointCpp/CacheDataModuleU.cpp
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable, 
// as such term is defined thereunder. Your use of this software constitutes 
// your acknowledgement of your agreement to the foregoing software license 
// and support agreement. 
//---------------------------------------------------------------------------

#include <System.JSON.hpp>
#include <System.Math.hpp>
#pragma hdrstop
#include "CacheDataModuleU.h"
#include <memory>
#include <map>
#include <vector>
// ---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma classgroup "FMX.Controls.TControl"
#pragma resource "*.dfm"

TCacheDataModule *CacheDataModule;

// ---------------------------------------------------------------------------
__fastcall TCacheDataModule::TCacheDataModule(TComponent* Owner)
    : TDataModule(Owner) {
    FCache = new TCache();
}

__fastcall TCacheDataModule::~TCacheDataModule(void) {
    delete FCache;

}

void __fastcall TCacheDataModule::SaveDeviceData(String ADevice,
    TDateTime ATime, TJSONObject * AData) {
    TItem* LItem = new TItem(ATime, ADevice, "", AData);
    FCache->Save(LItem);
}

void __fastcall TCacheDataModule::ClearDeviceData(String ADevice) {
    FCache->ClearDevice(ADevice);

}

bool __fastcall TCacheDataModule::TryGetRecentDeviceData(String ADevice,
    TDateTime &ATime, TJSONObject * AData) {
    TItem AItem;
    if (FCache->TryGetRecent(ADevice, AItem)) {
	ATime = AItem.Time;
	std::auto_ptr<TJSONObject>LJSON
	    (dynamic_cast<TJSONObject*>(TJSONObject::ParseJSONValue
	    (AItem.Value)));
	if (LJSON.get() != NULL) {

	    for (int i = 0; i < LJSON->Count; i++) {
		AData->AddPair
		    (dynamic_cast<TJSONPair*>(LJSON->Pairs[i]->Clone()));
	    }
	}
	return true;
    }
    else
	return false;
}

typedef System::DynamicArray<double>TDoublesArray;

class Stats {
public:
    TDateTime FTime;
    bool FFound;

    std::map<String, TDoublesArray>FMap;

    Stats::Stats() {
	FFound = false;
    }

    void __fastcall OnItem(const TItem &AItem, bool &ADone) {
	if (!FFound)
	    FTime = AItem.Time;
	FFound = true; // Found
	std::auto_ptr<TJSONObject>LJSON
	    (dynamic_cast<TJSONObject*>(TJSONObject::ParseJSONValue
	    (AItem.Value)));
	assert(LJSON.get() != NULL);
	for (int i = 0; i < LJSON->Count; i++) {
	    TJSONNumber * LJSONNumber =
		(dynamic_cast<TJSONNumber*>(LJSON->Pairs[i]->JsonValue));
	    if (LJSONNumber != NULL) {
		String key = LJSON->Pairs[i]->JsonString->Value();
		TDoublesArray &array = FMap[key];
		array.set_length(array.Length + 1);
		array[array.Length - 1] = LJSONNumber->AsDouble;
	    }
	}
    }
};

bool __fastcall TCacheDataModule::TryGetDeviceDataStats(String ADevice,
    TDateTime &ATime, TJSONObject *AData) {
    std::auto_ptr<Stats>LStats(new Stats());

    FCache->EnumItems(ADevice, LStats->OnItem);
    if (LStats->FFound) {
	ATime = LStats->FTime;
	std::map<String, TDoublesArray>::iterator it;
	for (it = LStats->FMap.begin(); it != LStats->FMap.end(); ++it) {
	    double LMean;
	    double LStdDev;
	    TDoublesArray &LList = it->second;
	    System::Math::MeanAndStdDev(&LList[0], LList.Length - 1, LMean,
		LStdDev);
	    std::auto_ptr<TJSONObject>LJSONObject(new TJSONObject());
	    LJSONObject->AddPair("last", new TJSONNumber(LList[0]));
	    LJSONObject->AddPair("mean", new TJSONNumber(LMean));
	    LJSONObject->AddPair("stddev", new TJSONNumber(LStdDev));
	    String key = it->first;
	    AData->AddPair(key, LJSONObject.release());
	}

    }
    return LStats->FFound;

}

bool __fastcall TCache::TryGetRecent(String ADevice, TItem &AItem) {
    std::vector<TItem>&v = *FList;
    std::vector<TItem>::iterator it;
    for (it = v.begin(); it < v.end(); it++) {
	if (ADevice == it->Device) {
	    AItem = *it;
	    return true;
	}
    }
    return false;
}

bool __fastcall TCache::EnumItems(String ADevice, TItemsEvent AOnItem) {
    std::vector<TItem>&v = *FList;
    std::vector<TItem>::iterator it;
    for (it = v.begin(); it < v.end(); it++) {
	if (ADevice == it->Device) {
	    bool LDone = false;
	    AOnItem(*it, LDone);
	    if (LDone) {
		return true;
	    }
	}
    }
    return false;
}

void __fastcall TCache::ClearDevice(String ADevice) {
    std::vector<TItem>&v = *FList;
    std::vector<TItem>::iterator it;
    for (it = v.begin(); it < v.end(); it++) {
	if (ADevice == it->Device) {
	    std::vector<TItem>::iterator del = it;
	    it++;
	    v.erase(del);
	}
    }
}

void __fastcall TCache::Save(TItem * AItem) {
    bool append = true;
    std::vector<TItem>&v = *FList;
    std::vector<TItem>::iterator it;
    for (it = v.begin(); it < v.end(); it++) {
	if (AItem->Time > it->Time) {
	    append = false;
	    v.insert(it, *AItem);
	    break;
	}
    }
    if (append) {
	v.push_back(*AItem);
    }
}

// ---------------------------------------------------------------------------

File: Embarcadero/Rio/Object Pascal/ThingPoint ThingConnect IoT Demo/Thingpoint/CacheDataModuleU.pas
//---------------------------------------------------------------------------
// Copyright (c) 2016 Embarcadero Technologies, Inc. All rights reserved.  
//
// This software is the copyrighted property of Embarcadero Technologies, Inc. 
// ("Embarcadero") and its licensors. You may only use this software if you 
// are an authorized licensee of Delphi, C++Builder or RAD Studio 
// (the "Embarcadero Products").  This software is subject to Embarcadero's 
// standard software license and support agreement that accompanied your 
// purchase of the Embarcadero Products and is considered a Redistributable, 
// as such term is defined thereunder. Your use of this software constitutes 
// your acknowledgement of your agreement to the foregoing software license 
// and support agreement. 
//---------------------------------------------------------------------------
unit CacheDataModuleU;

interface

uses
  System.SysUtils, System.Classes, System.Generics.Collections,
  System.JSON;

type

  TCache = class
  public type
    TItem = record
    private
      FTime: TDateTime;
      FDevice: string;
      FValue: string;
      FPerson: string;
    public
      constructor Create(ATime: TDateTime; const ADevice, APerson: string;
        const AValue: TJSONObject);
      property Person: string read FPerson;
      property Device: string read FDevice;
      property Time: TDateTime read FTime;
      property Value: string read FValue;
    end;
    TEnumCallback = reference to function(const AItem: TItem): Boolean;
  private
    FList: TThreadList<TItem>;
  public
    constructor Create;
    destructor Destroy; override;
    procedure ClearDevice(const ADevice: string);
    procedure Save(const AItem: TItem);
    function EnumValues(const ACallback: TEnumCallback): Boolean;
  end;

  TCacheDataModule = class(TDataModule)
  private
    FCache: TCache;
  public
    { Public declarations }
    constructor Create(AOwner: TComponent); override;
    destructor Destroy; override;
    procedure SaveDeviceData(const ADevice: string; ATime: TDateTime; AData: TJSONObject);
    procedure ClearDeviceData(const ADevice: string);
    function TryGetRecentDeviceData(const ADevice: string; out ATime: TDateTime; const AData: TJSONObject): Boolean;
    function TryGetDeviceDataStats(const ADevice: string; out ATime: TDateTime; const AData: TJSONObject): Boolean;
  end;

var
  CacheDataModule: TCacheDataModule;

implementation

uses System.Math;

{%CLASSGROUP 'FMX.Controls.TControl'}

{$R *.dfm}

{ TCacheDataModule }


constructor TCacheDataModule.Create(AOwner: TComponent);
begin
  inherited;
  FCache := TCache.Create;

end;

destructor TCacheDataModule.Destroy;
begin
  FCache.Free;
  inherited;
end;


procedure TCacheDataModule.SaveDeviceData(const ADevice: string; ATime: TDateTime; AData: TJSONObject);
var
  LItem: TCache.TItem;
begin
  LItem := TCache.TItem.Create(ATime, ADevice, '', AData);
  FCache.Save(LItem);
end;

procedure TCacheDataModule.ClearDeviceData(const ADevice: string);
begin
  FCache.ClearDevice(ADevice);
end;

function TCacheDataModule.TryGetRecentDeviceData(const ADevice: string; out ATime: TDateTime; const AData: TJSONObject): Boolean;
var
  LValue: TCache.TItem;
  LJSON: TJSONObject;
  LPair: TJSONPair;
begin
  Result := FCache.EnumValues(
    function(const AValue: TCache.TItem): Boolean
    begin
      Result := AValue.Device = ADevice;
      if Result then
        LValue := AValue;
    end);

  if Result then
  begin
    ATime := LValue.Time;
    LJSON := TJSONObject.ParseJSONValue(LValue.Value) as TJSONObject;
    try
      Assert(LJSON <> nil);
      for LPair in LJSON do
        AData.AddPair(LPair.Clone as TJSONPair);
    finally
      LJSON.Free;
    end;
  end;
end;


// TODO: Cache most recent calculation and time
function TCacheDataModule.TryGetDeviceDataStats(const ADevice: string; out ATime: TDateTime; const AData: TJSONObject): Boolean;
var
  LJSONObject: TJSONObject;
  LDataLists: TDictionary<string, TList<Double>>;
  LTime: TDateTime;
  LPair: TPair<string, TList<Double>>;
  LMean: Double;
  LStdDev: Double;
  LList: TList<Double>;
  LResult: Boolean;
begin
  LDataLists := TObjectDictionary<string, TList<Double>>.Create([doOwnsValues]);
  try
    FCache.EnumValues(
      function(const AValue: TCache.TItem): Boolean
      var
        LJSONPair: TJSONPair;
      begin
        Result := False; // Keep going
        if AValue.Device = ADevice then
        begin
          if not LResult then
            LTime := AValue.Time;
          LResult := True; // Found
          LJSONObject := TJSONObject.ParseJSONValue(AValue.Value) as TJSONObject;
          try
            Assert(LJSONObject <> nil);
            for LJSONPair in LJSONObject do
            begin
              if LJSONPair.JsonValue is TJSONNumber then
              begin
               if not LDataLists.TryGetValue(LJSONPair.JsonString.Value, LList) then
               begin
                 LList := TList<Double>.Create;
                 LDataLists.Add(LJSONPair.JsonString.Value, LList);
               end;
               LList.Add(TJSONNumber(LJSONPair.JsonValue).AsDouble)
              end;
            end;
          finally
            LJSONObject.Free;
          end;
        end;
      end);

    if LResult then
    begin
      ATime := LTime;
      for LPair in  LDataLists do
      begin
        LMean := 0;
        LStdDev := 0;
        LList := LPair.Value;
        System.Math.MeanAndStdDev(LList.ToArray, LMean, LStdDev);
        LJSONObject := TJSONObject.Create;
        AData.AddPair(LPair.Key, LJSONObject);
        LJSONObject.AddPair('last', TJSONNumber.Create(LPair.Value[0]));
        LJSONObject.AddPair('mean', TJSONNumber.Create(LMean));
        LJSONObject.AddPair('stddev', TJSONNumber.Create(LStdDev));
      end;
    end;
  finally
    LDataLists.Free;
  end;
  Result := LResult;
end;

{ TCache }

constructor TCache.Create;
begin
  inherited;
  FList := TThreadList<TCache.TItem>.Create;

end;

destructor TCache.Destroy;
begin
  FList.Free;
  inherited;
end;

function TCache.EnumValues(const ACallback: TEnumCallback): Boolean;
var
  I: Integer;
  LList: TList<TItem>;
  LResult: Boolean;
begin
  LResult := False;
  LList := FList.LockList;
  try
    for I := 0 to LList.Count - 1 do
    begin
      LResult := ACallback(LList[I]);
      if LResult then
        break;
    end;
  finally
    FList.UnlockList;
  end;
  Result := LResult;
end;

procedure TCache.Save(const AItem: TItem);
var
  I: Integer;
  Index: Integer;
  LList: TList<TItem>;
begin
  LList := FList.LockList;
  try
    Index := -1;
    for I := 0 to LList.Count - 1 do
    begin
      if AItem.Time > LList.Items[I].Time then
      begin
        Index := I;
        break;
      end;
    end;
    if Index >= 0 then
      LList.Insert(Index, AItem)
    else
      LList.Add(AItem);
  finally
    FList.UnlockList;
  end;
end;

procedure TCache.ClearDevice(const ADevice: string);
var
  I: Integer;
  LList: TList<TItem>;
begin
  LList := FList.LockList;
  try
    for I := LList.Count - 1 downto 0 do
      if LList.Items[I].Device = ADevice then
        LList.Delete(I);
  finally
    FList.UnlockList;
  end;
end;

{ TCache.TItem }

constructor TCache.TItem.Create(ATime: TDateTime; const ADevice, APerson: string;
  const AValue: TJSONObject);
begin
  FTime := ATime;
  FDevice := ADevice;
  FPerson := APerson;
  FValue := AValue.ToJson;
end;

end.
